test/e2e/cluster_upgrade_test.go

```diff
@@ -69,9 +83,122 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
 			// This test is run in CI in parallel with other tests. To keep the test duration reasonable
 			// the conformance tests are skipped.
 			SkipConformanceTests:     true,
-			ControlPlaneMachineCount: ptr.To[int64](3),
+			ControlPlaneMachineCount: ptr.To[int64](controlPlaneMachineCount),
 			WorkerMachineCount:       ptr.To[int64](1),
-			Flavor:                   ptr.To("topology"),
+			Flavor:                   ptr.To("kcp-pre-drain"),
+			PreWaitForControlPlaneToBeUpgraded: func(managementClusterProxy framework.ClusterProxy, workloadClusterNamespace, workloadClusterName string) {
+				log.Logf("Waiting for control-plane machines to have the upgraded Kubernetes version")
+
+				preDrainHook := "pre-drain.delete.hook.machine.cluster.x-k8s.io/kcp-ready-check"
+
+				cluster := &clusterv1.Cluster{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: workloadClusterNamespace,
+						Name:      workloadClusterName,
+					},
+				}
+				Expect(managementClusterProxy.GetClient().Get(ctx, client.ObjectKeyFromObject(cluster), cluster)).To(Succeed())
+
+				// This replaces the WaitForControlPlaneMachinesToBeUpgraded function to verify via a pre-drain hook
+				// that all static Pods, kube-proxy and all Nodes are becoming healthy before we let the upgrade
+				// process precede by removing the pre-drain hook.
+				// This captures cases where static Pods, kube-proxy and all Nodes would only become healthy after
+				// all control plane Machines have been upgraded.
+				Eventually(func() (int64, error) {
+					machines := framework.GetControlPlaneMachinesByCluster(ctx, framework.GetControlPlaneMachinesByClusterInput{
+						Lister:      managementClusterProxy.GetClient(),
+						ClusterName: cluster.Name,
+						Namespace:   cluster.Namespace,
+					})
+
+					// Collect information about:
+					// * how many control-plane machines already got upgradedAndHealthy
+					// * control-plane machines which are in deletion but waiting for the pre-drain hook
+					// * workload cluster nodes
+					// * kube-proxy pods
+					var upgradedAndHealthy int64
+					deletingMachines := []clusterv1.Machine{}
+					for _, m := range machines {
```

Comment:
@chrischdi Can you please open a separate PR to fix the description of the MachineNodeHealthyCondition?

```go
	// MachineNodeHealthyCondition provides info about the operational state of the Kubernetes node hosted on the machine by summarizing  node conditions.
	// If the conditions defined in a Kubernetes node (i.e., NodeReady, NodeMemoryPressure, NodeDiskPressure, NodePIDPressure, and NodeNetworkUnavailable) are in a healthy state, it will be set to True.
	MachineNodeHealthyCondition ConditionType = "NodeHealthy"
```
It's at least wrong because we never look at NodeNetworkUnavailable (didn't check if NodeNetworkUnavailable even exists)

Reply:
https://github.com/kubernetes-sigs/cluster-api/pull/11178